vllm/core/scheduler.py

```diff
@@ -395,6 +395,38 @@ def __init__(
         # for processing and deallocation by the free_finished_seq_groups()
         self._async_stopped: List[SequenceGroup] = []
 
+        # For chunked prefill, we allocate a set of "prefill slots" that
+        # each represent a sequence group that can be partially prefilled.
+        # Having multiple partial prefills in flight allows us to minimize TTFT
+        # and avoid decode starvation in cases where a single sequence group
+        # with a very large prompt blocks the queue for too many iterations.
+        self.num_prefill_slots = scheduler_config.num_prefill_slots
+        self.prefill_slots_running = 0
+        self.big_prefill_requests = 0
+        # Requests with more than (4% max context length) tokens to prefill
+        # are "big".
```

Comment:
The entire goal here is to not allow decode to be starved by the prefill phase blocking on long requests- this part of the PR description:
> A single very large prompt will block all other prompts from prefilling for many iterations. This can eventually starve decoding- for example a 130k token prompt with —max-num-batched-tokens=512 will take about 250 iterations to prefill, in which time the currently decoding sequences may all finish. Send a few of these requests at once and very quickly nothing will be decoding.

Just allowing concurrent partial prefills doesn't solve the problem by itself, because multiple long requests could still block up the prefill. So what we do is only allow a single long request to prefill, and allow smaller requests to be pulled from the waiting queue instead of more long ones